V2.0.3 tuenti

acampoh opened the pull request:
This PR tries to fix the galera HG management when you have multiple nodes shared between different galera HGs. Also it fixes the management of shunned nodes that were deleted from runtime, and now are moved properly to the offline HG so they can be recovered when they are back online automatically.

Also a couple of changes that i found useful for debugging, such as having the runtime sqlite db on file on debug compilations so you can check the contents in real time, and also allow any amount of monitor threads also in debug, so the behaviour of them can be debugged in a easier way.

Commit:
Fixed hostgroup generation with galera support. Added some ifdefs for debug purposes

Commit:
Fixed behaviour when a node is shunned / desynced can be automatically recovered.

Now if all nodes are shunned / desynced in a hostgroup we keep the one with highest weight as the "available" one

Commit:
Small refactor

pondix commented:
Automated message: PR pending admin approval for build testing

renecannao commented:
Hi @acampoh , thank you for the PR!

> This PR tries to fix the galera HG management when you have multiple nodes shared between different galera HGs. Also it fixes the management of shunned nodes that were deleted from runtime, and now are moved properly to the offline HG so they can be recovered when they are back online automatically.

Can you please describe what was the original issue, and how it is fixed?

>  Also a couple of changes that i found useful for debugging, such as having the runtime sqlite db on file on debug compilations so you can check the contents in real time

That was not needed. When compiled in debug mode, you can access the database from Admin, schema `myhgm`. Run this to confirm:
```
SHOW DATABASES;
SHOW TABLES FROM myhgm;
```

By the way, to be honest I am not completely happy with the design of Galera checks, because each node is checked independently and then Monitor takes action immediately while trying to not conflict with other checks.
For Aurora I am working on a different approach: for each cluster, a dedicated thread will check all the nodes of the cluster. This avoids conflicts.
The module isn't ready yet, work in progress, but I am mentioning because maybe you want to have a look at it.

acampoh commented:
Hello!

The issue we were having with the packaged version was that we found that we our setup (having 3 nodes shared between 3 HGs)  as sown below, proxysql was generating a bad runtime_mysql_servers setup while having all the nodes setup.

```
mysql_servers =
(
	{
		address="127.0.0.1"
		port=3310
		hostgroup=3307
		max_connections=200
		weight=10000000
		comment="test server"
	},
	{
		address="127.0.0.2"
		port=3311
		hostgroup=3307
		max_connections=200
		weight=1000
		comment="test server"
	},
	{
		address="127.0.0.3"
		port=3312
		hostgroup=3307
		max_connections=200
		weight=1
		comment="test server"
	},
	{
		address="127.0.0.2"
		port=3311
		hostgroup=3308
		max_connections=200
		weight=10000000
		comment="test server"
	},
	{
		address="127.0.0.3"
		port=3312
		hostgroup=3308
		max_connections=200
		weight=1000
		comment="test server"
	},
	{
		address="127.0.0.1"
		port=3310
		hostgroup=3308
		max_connections=200
		weight=1
		comment="test server"
	},
	{
		address="127.0.0.3"
		port=3312
		hostgroup=3309
		max_connections=200
		weight=10000000
		comment="test server"
	},
	{
		address="127.0.0.1"
		port=3310
		hostgroup=3309
		max_connections=200
		weight=1000
		comment="test server"
	},
	{
		address="127.0.0.2"
		port=3311
		hostgroup=3309
		max_connections=200
		weight=1
		comment="test server"
	}
)

mysql_galera_hostgroups =
(
    {
	writer_hostgroup=3307
        backup_writer_hostgroup=13307
        reader_hostgroup=23307
        offline_hostgroup=33307
        active=1
        max_writers=1
        writer_is_also_reader=1
        comment="group1" 
    },
    {
	writer_hostgroup=3308
        backup_writer_hostgroup=13308
        reader_hostgroup=23308
        offline_hostgroup=33308
        active=1
        max_writers=1
        writer_is_also_reader=1
        comment="group2" 
    },
    {
	writer_hostgroup=3309
        backup_writer_hostgroup=13309
        reader_hostgroup=23309
        offline_hostgroup=33309
        active=1
        max_writers=1
        writer_is_also_reader=1
        comment="group3" 
    }
)
```

as proxysql was looking for a host:port in a HG != writer_hg that was including all HGs setup and not only the ones related with that galera HG.

Also once the happy case was working for us we saw that for desync / shunned nodes, they were deleted from runtime_mysql_servers instead of moving them to the offline HG so it was unable to recover the nodes after recovering the node. Now shunned / offline_soft nodes will be moved to offline_hg so when they are up again they will be moved to the corresponding HG (writer / reader / backup) as set up.

About moving the db from inmemory to file i tried to connect to it several times but i wasn't able to explore the content. I'll try again and remove that part in case of success.

I'll check the Aurora part so i can have a good idea of how that will work (sounds great what you said) but currently this the best approach i found without modifying the behaviour of the monitor

acampoh commented:
After some testing in a real environment with traffic and playing with it a bit with the node status i have found some corner cases not covered in this PR. I'll find them and update this

renecannao commented:
I have some feedback from my end too.
In the context of Galera, a writer_hostgroup represent a cluster.
Although it is technically possible for a server to be part of multiple hostgroups (and this is quite common), the Galera module wasn't designed to have a server part of multiple clusters. Thus i understand why your case doesn't work.
But I want to to ask: why do you have such setup? Why do you have a single server part of multiple clusters? Am I missing something?

emottola commented:
Hi René, I'll answer your question about our setup.

First I need to thank you and the Proxysql team for having developed a such powerful and helpful tool. We're really enjoying it and using it on production since some months in our large multi-DCs Galera-based MySQL installation.
Since we're widely using and testing it we found some limitation that we need to address for moving to Proxysql 2.0 with its native galera support.

In our setup we split the load (read+write traffic together) by software configuration across all the nodes in the clusters, that allows use to not over/under use any of the nodes.
We archive this in two ways:
1) A schema-level configuration: all traffic for one schemas goes on the same node, taking into account we manage big boxes with multiple database schemas deployed.
2) A table-level configuration using logical sharding: for the target PK values our internal software library calculates the right table numeric suffix (so the tables are like _table1_ or _table2_ or _tableX_)  and then check/use the appropriate database configuration for that specific [range of] table.

Each Galera Cluster has its own Proxysql deployment (multiple pods) that manage the traffic in front of it with multiple HGs configured.
Each Host Group has one of the server configured as "primary", the remaining ones as "secondary".

When all the nodes are online, each of them takes only its own specific set of traffic (schemas+tables), that's why I called "primary", despite the technical definition should be "preferred node".
If this node goes down the traffic is routed to **one** of the remaining available nodes.
This is the reason why every single server is part of multiple HGs, because each server is the preferred one in one HGs and takes almost all the traffic, in the remaining HGs the same server it's configured to be used only in case of loss of the preferred.

In order to define the preferred node in a given HG where we have 3 ONLINE nodes, using proxysql 1.4.x we're playing with server weights like assigning value of 1000000 to the 1st node, 1000 for 2nd node and 1 for 3rd node.
This is because we don't like/adopt the logic of the existing galera checker script and don't want to wait for scheduler/checker script actions when a node goes down.
So in every HG we always have 3 or more available nodes but we send traffic "mostly" to only one of them, since the weight-ratio traffic distribution it's not super precise, but we don't care.
Our goal is to have proxysql 2.0 fully natively capable of managing all the Galera cluster possible server/node status while having the same server in multiple HGs if needed.

As a context information, we have a 1:1 ration between read and write traffic, so we don't need to use one node for writes and multiple ones for reads, we "only" need extremely high availability and fast traffic switching.
Once a server has a problem (goes down or become desync or whatever) we need to move traffic **as soon as possible** in terms of sub-seconds, and IMHO using proxysql internal monitor is the best solution for this job, so let's improve it!

Please let me know if our uses case and our way of using proxysql makes sense to you and please do not hesitate to ask, comment or suggest anything.

I'm of course open to any deeper discussion about it, you can poke me on Slack directly  :)

acampoh commented:
Hello Rene,

I have found the reason of the weird flapping of the nodes when proxysql has load. in **MySQL_Monitor.cpp::monitor_galera_thread()** sometimes the query you make for getting the state of nodes, sometimes is getting partial results (3 columns instead of 7), so after that is making decissions of which nodes to put offline without the proper info.

I'm not sure of the reason of this, maybe a bug in that mariadb lib version, or a dangling pointer... I checked the network packaged with wireshark and everything looked ok to me. I tried changing the goto by a return, and it's working fine, but i think that's not a fix, just a workaround. 

I'll try to findout what happens and let you know.

renecannao commented:
> I'm not sure of the reason of this, maybe a bug in that mariadb lib version, or a dangling pointer...

I literally spent 2 weeks on this! :-)
Fixed in 2.0.4

renecannao commented:
To give more context, I spent a lot of time thinking the same: "maybe a bug in that mariadb lib version, or a dangling pointer" .
At the end, it turned out that on some circumstances (mostly due to timeouts) connections were incorrectly returning into Monitor's connection pool, and the next thread using them were performing dirty reads

acampoh commented:
that's great! I'll rebase my changes and test this again.

Thanks a lot for your quick response

acampoh commented:
Just tested and it's working for me as expected. There is no more flapping with the node status. \o/

Commit:
Fixed .deb build for debian images

renecannao commented:
> Just tested and it's working for me as expected. There is no more flapping with the node status. \o/

Good news! :)
We didn't have the time to formally test it against 2.0.4 (to be released today), thus we will review the PR for the next version.
Thanks

Commit:
Finally fixed all stuff for full galera support.

- Fixed galera status queries, which made the mysql_server table not to recover properly after full cluster recovery.
- Changed wait_for_mysql calls by usleeps as they were causing false timeouts. Now timeouts are managed by mysql_query_cont on Mysql_Monitor as also is able to manage timeouts.
- Recovered check for weird return values from mysql conns as doesn't hurt and can be useful.
- small refactor (remove comments / indented some code)

acampoh committed:
Merge branch 'v2.0.4' into v2.0.3-tuenti

pondix commented:
Hi @acampoh - would it be possible to please make a PR for this patch against 2.0.5? I'll then run the automated build tests to verify

acampoh commented:
Sure! I'll do it tomorrow morning

acampoh commented:
Done! It seems the patch still goes clean with v2.0.5

pondix commented:
Thanks!

pondix commented:
test this please

acampoh commented:
Hello, i was reviewing the state of this and i'm not sure if i still have to do something to get it approved or it's ok as it is. Let me know if there is something else i can do :) 

Just as a side note, we have been using this version in production (without galera_cluster config) for several weeks and with the proper config since this morning.

acampoh commented:
Also after applying galera cluster config and removing the old checker from the scheduler we have reduced by 3 the cpu usage (i suppose due to having 3 nodes and a scheduler for each node)

![image](https://user-images.githubusercontent.com/1070678/60006454-bf35fa00-9670-11e9-8612-0019467866de.png)

renecannao commented:
Hi @acampoh .
Thank you for the update.
This is blocked on our end, and we are trying to prioritize it.

acampoh commented:
thanks a lot for your update, there is no rush from our side. If you need us to change something related, let us know and we'll update the PR.

renecannao commented on lib/MySQL_HostGroups_Manager.cpp:
```diff
@@ -969,7 +969,7 @@ MySQL_HostGroups_Manager::MySQL_HostGroups_Manager() {
 	mydb=new SQLite3DB();
 #ifdef DEBUG
-	mydb->open((char *)"file:mem_mydb?mode=memory&cache=shared", SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX);
+	mydb->open((char *)"/tmp/proxysql/inmemory.db", SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX);
```
This is not needed. Tables are all accessible using `myhgm` schema.
```
Admin> SELECT @@version;
+-------------------------+
| @@version               |
+-------------------------+
| 2.0.5-36-g7e33124_DEBUG |
+-------------------------+
1 row in set (0.00 sec)

Admin> SHOW DATABASES;
+-----+---------------+------------------------------------------------+
| seq | name          | file                                           |
+-----+---------------+------------------------------------------------+
| 0   | main          |                                                |
| 2   | disk          | /home/rcannao/proxysql/src/./proxysql.db       |
| 3   | stats         |                                                |
| 4   | monitor       |                                                |
| 5   | stats_history | /home/rcannao/proxysql/src/./proxysql_stats.db |
| 6   | myhgm         |                                                |
+-----+---------------+------------------------------------------------+
6 rows in set (0.00 sec)

Admin> SHOW TABLES FROM myhgm;
+------------------------------------+
| tables                             |
+------------------------------------+
| mysql_aws_aurora_hostgroups        |
| mysql_galera_hostgroups            |
| mysql_group_replication_hostgroups |
| mysql_replication_hostgroups       |
| mysql_servers                      |
| mysql_servers_incoming             |
+------------------------------------+
6 rows in set (0.00 sec)
```

renecannao replied:
Reverted

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -745,7 +750,6 @@ void * monitor_ping_thread(void *arg) {
 	mmsd->interr=0; // reset the value
 	mmsd->async_exit_status=mysql_ping_start(&mmsd->interr,mmsd->mysql);
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
You are replacing `wait_for_mysql()` with an `usleep(10000)` .
In term of maximum waiting time is the same, as `wait_for_mysql()` calls `poll()` with a timeout of 10ms.
Although the use of `wait_for_mysql()` is better than `usleep()` because it uses a timeout instead of a fixed pause (10ms), the use of `wait_for_mysql()` becomes mandatory when its returned value is then used as an argument by async API calls like `mysql_ping_cont` and `mysql_query_cont` .

renecannao replied:
Reverting all the changes related to `wait_for_mysql()` and `usleep()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -757,6 +761,7 @@ void * monitor_ping_thread(void *arg) {
 		if ((mmsd->async_exit_status & MYSQL_WAIT_TIMEOUT) == 0) {
 			mmsd->async_exit_status=mysql_ping_cont(&mmsd->interr, mmsd->mysql, mmsd->async_exit_status);
 		}
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -862,7 +867,6 @@ bool MySQL_Monitor_State_Data::set_wait_timeout() {
 	t1=monotonic_time();
 	async_exit_status=mysql_query_start(&interr,mysql,query);
 	while (async_exit_status) {
-		async_exit_status=wait_for_mysql(mysql, async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -874,6 +878,7 @@ bool MySQL_Monitor_State_Data::set_wait_timeout() {
 		if ((async_exit_status & MYSQL_WAIT_TIMEOUT) == 0) {
 			async_exit_status=mysql_query_cont(&interr, mysql, async_exit_status);
 		}
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -987,6 +991,7 @@ void * monitor_read_only_thread(void *arg) {
 		if ((mmsd->async_exit_status & MYSQL_WAIT_TIMEOUT) == 0) {
 			mmsd->async_exit_status=mysql_query_cont(&mmsd->interr, mmsd->mysql, mmsd->async_exit_status);
 		}
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -968,7 +973,6 @@ void * monitor_read_only_thread(void *arg) {
 		}
 	}
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -995,7 +1000,6 @@ void * monitor_read_only_thread(void *arg) {
 	}
 	mmsd->async_exit_status=mysql_store_result_start(&mmsd->result,mmsd->mysql);
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1009,6 +1013,8 @@ void * monitor_read_only_thread(void *arg) {
 			mmsd->async_exit_status=mysql_store_result_cont(&mmsd->result, mmsd->mysql, mmsd->async_exit_status);
 		}
+
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1217,6 +1222,7 @@ void * monitor_group_replication_thread(void *arg) {
 		if ((mmsd->async_exit_status & MYSQL_WAIT_TIMEOUT) == 0) {
 			mmsd->async_exit_status=mysql_query_cont(&mmsd->interr, mmsd->mysql, mmsd->async_exit_status);
 		}
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1198,7 +1204,6 @@ void * monitor_group_replication_thread(void *arg) {
 	mmsd->interr=0; // reset the value
 	mmsd->async_exit_status=mysql_query_start(&mmsd->interr,mmsd->mysql,"SELECT viable_candidate,read_only,transactions_behind FROM sys.gr_member_routing_candidate_status");
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1238,6 +1243,8 @@ void * monitor_group_replication_thread(void *arg) {
 			mmsd->async_exit_status=mysql_store_result_cont(&mmsd->result, mmsd->mysql, mmsd->async_exit_status);
 		}
+
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1225,7 +1231,6 @@ void * monitor_group_replication_thread(void *arg) {
 	}
 	mmsd->async_exit_status=mysql_store_result_start(&mmsd->result,mmsd->mysql);
 	while (mmsd->async_exit_status && ((mmsd->async_exit_status & MYSQL_WAIT_TIMEOUT) == 0)) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1517,17 +1523,20 @@ void * monitor_galera_thread(void *arg) {
+	mmsd->t1=monotonic_time();
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
+		//mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1538,10 +1547,11 @@ void * monitor_galera_thread(void *arg) {
 			mmsd->async_exit_status=mysql_query_cont(&mmsd->interr, mmsd->mysql, mmsd->async_exit_status);
 		}
+
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1538,10 +1547,11 @@ void * monitor_galera_thread(void *arg) {
 	}
 	mmsd->async_exit_status=mysql_store_result_start(&mmsd->result,mmsd->mysql);
 	while (mmsd->async_exit_status && ((mmsd->async_exit_status & MYSQL_WAIT_TIMEOUT) == 0)) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1554,6 +1564,8 @@ void * monitor_galera_thread(void *arg) {
 			mmsd->async_exit_status=mysql_store_result_cont(&mmsd->result, mmsd->mysql, mmsd->async_exit_status);
 		}
+
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1892,14 +1899,11 @@ void * monitor_replication_lag_thread(void *arg) {
-		//	mmsd->mysql_error_msg=strdup("timeout check");
-		//	goto __exit_monitor_replication_lag_thread;
 		}
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1881,7 +1889,6 @@ void * monitor_replication_lag_thread(void *arg) {
 		mmsd->async_exit_status=mysql_query_start(&mmsd->interr,mmsd->mysql,"SHOW SLAVE STATUS");
 	}
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1892,14 +1899,11 @@ void * monitor_replication_lag_thread(void *arg) {
 	}
 	mmsd->async_exit_status=mysql_store_result_start(&mmsd->result,mmsd->mysql);
 	while (mmsd->async_exit_status && ((mmsd->async_exit_status & MYSQL_WAIT_TIMEOUT) == 0)) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -1910,10 +1914,9 @@ void * monitor_replication_lag_thread(void *arg) {
-		//	goto __exit_monitor_replication_lag_thread;
 		}
+
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -3923,10 +3925,11 @@ void * monitor_AWS_Aurora_thread_HG(void *arg) {
 			mmsd->async_exit_status=mysql_query_cont(&mmsd->interr, mmsd->mysql, mmsd->async_exit_status);
 		}
+
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -3910,7 +3913,6 @@ void * monitor_AWS_Aurora_thread_HG(void *arg) {
 	mmsd->async_exit_status = mysql_query_start(&mmsd->interr, mmsd->mysql, "SELECT SERVER_ID, SESSION_ID, LAST_UPDATE_TIMESTAMP, REPLICA_LAG_IN_MILLISECONDS, CPU FROM INFORMATION_SCHEMA.REPLICA_HOST_STATUS ORDER BY SERVER_ID");
 #endif // TEST_AURORA
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -4352,6 +4356,8 @@ void * monitor_AWS_Aurora_thread(void *arg) {
 			mmsd->async_exit_status=mysql_store_result_cont(&mmsd->result, mmsd->mysql, mmsd->async_exit_status);
 		}
+
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -4336,10 +4340,10 @@ void * monitor_AWS_Aurora_thread(void *arg) {
 		if ((mmsd->async_exit_status & MYSQL_WAIT_TIMEOUT) == 0) {
 			mmsd->async_exit_status=mysql_query_cont(&mmsd->interr, mmsd->mysql, mmsd->async_exit_status);
 		}
+		usleep(10000);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -4336,10 +4340,10 @@ void * monitor_AWS_Aurora_thread(void *arg) {
 	}
 	mmsd->async_exit_status=mysql_store_result_start(&mmsd->result,mmsd->mysql);
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -4323,7 +4328,6 @@ void * monitor_AWS_Aurora_thread(void *arg) {
 	mmsd->async_exit_status = mysql_query_start(&mmsd->interr, mmsd->mysql, "SELECT SERVER_ID, SESSION_ID, LAST_UPDATE_TIMESTAMP, REPLICA_LAG_IN_MILLISECONDS, CPU FROM INFORMATION_SCHEMA.REPLICA_HOST_STATUS ORDER BY SERVER_ID");
 #endif // TEST_AURORA
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -3923,10 +3925,11 @@ void * monitor_AWS_Aurora_thread_HG(void *arg) {
 	}
 	mmsd->async_exit_status=mysql_store_result_start(&mmsd->result,mmsd->mysql);
 	while (mmsd->async_exit_status) {
-		mmsd->async_exit_status=wait_for_mysql(mmsd->mysql, mmsd->async_exit_status);
```
See review comment about `wait_for_mysql()`

renecannao commented on lib/MySQL_Monitor.cpp:
```diff
@@ -565,7 +565,12 @@ MySQL_Monitor::MySQL_Monitor() {
 	monitordb->execute("CREATE INDEX IF NOT EXISTS idx_aws_aurora_log_time_start ON mysql_server_aws_aurora_log (time_start_us)");
 
+#ifdef DEBUG
+	num_threads=1;
```
Is this relevant?
The initial value of `num_threads` only defines the number of threads during initialization, but this can go up to `mysql-monitor_threads_queue_maxsize` (default 128, max 1024).
I would remove this part.

acampoh replied:
this is very useful when debugging the code (at least using eclipse) as with multiple threads the debugger jumps from one thread to another making impossible to know what's happening. Being able to have only 1 thread running at a time makes much easier to debug the monitor module

renecannao replied:
I don't know if Eclipse supports it, but `gdb` has [scheduler-locking](https://sourceware.org/gdb/onlinedocs/gdb/All_002dStop-Mode.html#All_002dStop-Mode)

renecannao reviewed (commented):
Changing the base of this PR to a new branch, v2.0.5-2028 .
I will make some change before merging into 2.0.5.

Thanks
